doc/design.md

```diff
@@ -0,0 +1,928 @@
+  At a later point, in case we have convinced ourselves that the single-precision version is always sufficient, the default could be changed.
+
+It is not entirely clear how much development overhead this would cause for code depending on this library.
+More code is likely to require templating for the precision (`double` and `single`).
+However, the development cost for this library itself would be rather minor, so there is very little risk involved in supporting it from the beginning.
+
+
+## <a name="implementation"></a>Implementation
+
+
+### <a name="implementation-goals-and-non-goals"></a>Goals and non-goals
+
+##### Goals in this implementation phase
+
+1. Provide a simple, lightweight, functional and well-tested library that can replace and enhance most of the workspaces types existing in Mantid.
+1. Reach a point where this library (in combination with some supporting libraries) is useful in practice.
+   We define useful as:
+   - Complete low-level functionality in C++ and Python.
+   - `numpy` interoperability.
+   - Visualization using a slice viewer.
+   - Visualization using a multi-dimensional table viewer.
+   - Instrument visualization using the Mantid instrument viewer.
+   - Can leverage existing Mantid code such as `DataHandling::LoadEventNexus` by a subsequent conversion into a `Dataset`.
+1. Demonstrate usefulness on at least two real examples, ideally bringing them to a production ready stage.
+   Candidates are:
+   - Constant-wavelength (reactor) workflows, which currently use `API::MatrixWorkspace` with length-1 histograms as a workaround.
+   - Imaging workflows, which would benefit from presenting data as a stack of images rather than a list of histograms, i.e., the dimension order would be the opposite of what we are used to in `API::MatrixWorkspace`.
+   - Workflow with parameter scan or polarization analysis to demonstrate new capabilities of extra dimensions and/or multiple data variables.
+1. If applicable, demonstrate performance gain over existing Mantid workflow.
+
+##### Non-goals in this implementation phase
+
+1. It is a non-goal to provide a drop-in replacement for all existing workspaces.
+1. It is a non-goal to convert all 970 algorithms to use `Dataset` instead of `API::Workspace`.
+1. It is a non-goal to provide support for all visualization widgets.
+
+
+### <a name="implementation-milestones"></a>Milestones
+
+*To do:
+This needs a lot of work.
+It is quite unclear what belongs to which milestone, the idea is the have something more usable at every step.
+Also, do we have any ideas that can ultimately avoid a hard and risky "drop all the old stuff" step?
+*
+
+##### Milestone 1
+
+- Mostly useable in C++ and Python but basics not 100% complete.
+- Guerilla usability testing at user/dev workshop 2019.
+
+##### Milestone 2
+
+- Basics complete.
+- I/O, or at least converters from Mantid workspaces.
```

Comment:
Important point about `converting` but no type `compatibility`. Do we bring reasoning over from confluence notes on why we want to avoid that.

Reply:
I had touched on that with the first of the non-goals above, but will make it more explicit.